### ISSUE TITLE
Vim-command output disappears when a terminal buffer gets updated

### ISSUE THREAD

**Issue by critiqjo** [neovim/neovim#2503]
**Issue:** Run a script in terminal which outputs something periodically. Now if you execute `:version` in vim, it shows up first, then gets blacked out as soon as terminal screen is updated (the cursor remains in the command line). This happens even if you are in a new tab where you can't see the terminal.

**Steps to reproduce:** Run this shell script in terminal:
```sh
#!/bin/sh
while true; do
  echo "Test"
  sleep 1
done
```
Now, get back to normal mode, go to a new tab, and run `:version`... Also, tried this in `nvim -u NONE`

System details:
```
NVIM 0.0.0-alpha+201504230827 (compiled Apr 23 2015 17:47:13)
Commit: 04e098fc3c485eaa6d323e9121e9c81215d94a87
Optional features included (+) or not (-): +acl  +iconv
Operating System: Arch Linux
```

**Comment by justinmk** [comment]
This is the same root cause as https://github.com/neovim/neovim/issues/3382, I think.

**Comment by critiqjo** [comment]
There is another bug, probably related to this. Launch Neovim and open two buffers. Add 4 lines to second buffer. Now, execute this script:

```python
from neovim, time
nvim = neovim.attach('socket', path='/tmp/nvim')
buffer = nvim.buffers[0]
t = 0
while True:
  if t == 0:
    a = 1
  elif t == 4:
    a = -1
  t += a
  buffer[:] = ['line' for i in range(t+1)]
  time.sleep(0.5)
```

Bug 1:  The cursor will start moving up and down (even in insert mode) (depending on the difference in number of lines of the updated buffer)
Bug 2: Pressing `:` to go to command line will terminate the python script with an exception -- `NvimError: Vim:E315: ml_get: invalid lnum: XX`. (`XX` will be the line number where the cursor would have been positioned in the current buffer)

Note: If the first buffer is visible in a window or tab, there are no issues.

**Comment by critiqjo** [comment]
Fixed by #3413
